pandas/io/pytables.py

```diff
@@ -3020,12 +3050,50 @@ def set_info(self):
         """ update our table index info """
         self.attrs.info = self.info
 
+    def set_non_index_axes(self):
+        """ Write the axes to carrays """
+        def f(dim, flds):
+            name = "non_index_axes_%d" % dim
+            try:
+                self._handle.create_carray(self.attrs._v_node, name, obj=np.array(flds))
```

Comment:
I wasn't aware of that code.  My simple implementation of your suggestion sort of works, but actually raises different errors.  The root of the problem really seems to be non_index_axes, which converts proper index objects to arrays.  And at the moment lib.infer_dtype does the wrong thing for the test case - it says "datetime" for an array of Timestamps rather than datetime64 so the roundtrip of _unconvert_index(_convert_index()) is incorrect.   I'll have to investigate a bit more.